Exception on loop after agent call

```
    var creativeWriter = AgentsUtils.newCreativeWriter();
    var scorer = AgentsUtils.newStyleScorer();
    var editor = AgentsUtils.newStyleEditor();

    Predicate<AgenticScope> until = s -> s.readState("score", 0).doubleValue() >= 0.8;

    Map<String, Object> input =
            Map.of(
                    "topic", "fantasy",
                    "style", "fantasy",
                    "audience", "young adults");

    Workflow wf = workflow("loopFlow").agent(creativeWriter)
            .loop(until, scorer, editor).build();

    Object result;
    try (WorkflowApplication app = WorkflowApplication.builder().build()) {
      result = app.workflowDefinition(wf).instance(input).start().get().asJavaObject();
    } catch (Exception e) {
      throw new RuntimeException("Workflow execution failed", e);
    }

    System.out.println(result);
```


**What happened**:
```
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Optional.isPresent()" because the return value of "io.serverlessworkflow.api.types.func.ForTaskFunction.getForClass()" is null
	at io.serverlessworkflow.impl.executors.func.JavaForExecutorBuilder.collectionFilterObject(JavaForExecutorBuilder.java:70)
	at io.serverlessworkflow.impl.executors.func.JavaForExecutorBuilder.buildCollectionFilter(JavaForExecutorBuilder.java:65)
	at io.serverlessworkflow.impl.executors.ForExecutor$ForExecutorBuilder.<init>(ForExecutor.java:47)
	at io.serverlessworkflow.impl.executors.func.JavaForExecutorBuilder.<init>(JavaForExecutorBuilder.java:37)
	at io.serverlessworkflow.impl.executors.func.JavaTaskExecutorFactory.getTaskExecutor(JavaTaskExecutorFactory.java:30)
	at io.serverlessworkflow.impl.executors.TaskExecutorHelper.createExecutorBuilderList(TaskExecutorHelper.java:93)
	at io.serverlessworkflow.impl.executors.TaskExecutorHelper.createExecutorList(TaskExecutorHelper.java:65)
	at io.serverlessworkflow.impl.WorkflowDefinition.<init>(WorkflowDefinition.java:59)
	at io.serverlessworkflow.impl.WorkflowDefinition.of(WorkflowDefinition.java:69)
	at io.serverlessworkflow.impl.WorkflowDefinition.of(WorkflowDefinition.java:64)
	at io.serverlessworkflow.impl.WorkflowApplication.lambda$workflowDefinition$0(WorkflowApplication.java:257)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at io.serverlessworkflow.impl.WorkflowApplication.workflowDefinition(WorkflowApplication.java:256)
	at io.serverlessworkflow.fluent.agentic.langchain4j.WorkflowAgentsIT.loopTest(WorkflowAgentsIT.java:142)
	... 3 more
```



**What you expected to happen**:

No Exception, resulted String returned from the workflow

--- a/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
+++ b/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
@@ -71,8 +71,7 @@ public class JavaForExecutorBuilder extends ForExecutorBuilder {
   private Object collectionFilterObject(ForTaskFunction taskFunctions) {
     return taskFunctions
         .getForClass()
-        .<Object>map(
-            forClass -> (Object) new TypedFunction(taskFunctions.getCollection(), (Class) forClass))
+        .map(forClass -> new TypedFunction(taskFunctions.getCollection(), forClass))
         .orElse(taskFunctions.getCollection());
   }
 }

--- a/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
+++ b/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
@@ -69,7 +69,7 @@ public class JavaForExecutorBuilder extends ForExecutorBuilder {
   }
 
   private Object collectionFilterObject(ForTaskFunction taskFunctions) {
-    return taskFunctions.getForClass().isPresent()
+    return taskFunctions.getForClass() != null && taskFunctions.getForClass().isPresent()
         ? new TypedFunction(
             taskFunctions.getCollection(), taskFunctions.getForClass().orElseThrow())
         : taskFunctions.getCollection();

--- a/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
+++ b/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
@@ -69,9 +69,10 @@ public class JavaForExecutorBuilder extends ForExecutorBuilder {
   }
 
   private Object collectionFilterObject(ForTaskFunction taskFunctions) {
-    return taskFunctions.getForClass() != null && taskFunctions.getForClass().isPresent()
-        ? new TypedFunction(
-            taskFunctions.getCollection(), taskFunctions.getForClass().orElseThrow())
-        : taskFunctions.getCollection();
+    return taskFunctions
+        .getForClass()
+        .<Object>map(
+            forClass -> (Object) new TypedFunction(taskFunctions.getCollection(), (Class) forClass))
+        .orElse(taskFunctions.getCollection());
   }
 }

--- a/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
+++ b/experimental/lambda/src/main/java/io/serverlessworkflow/impl/executors/func/JavaForExecutorBuilder.java
@@ -71,7 +71,12 @@ public class JavaForExecutorBuilder extends ForExecutorBuilder {
   private Object collectionFilterObject(ForTaskFunction taskFunctions) {
     return taskFunctions
         .getForClass()
-        .map(forClass -> new TypedFunction(taskFunctions.getCollection(), forClass))
+        .<Object>map(forClass -> typedCollectionFunction(taskFunctions, forClass))
         .orElse(taskFunctions.getCollection());
+  }
+
+  @SuppressWarnings({"rawtypes", "unchecked"})
+  private Object typedCollectionFunction(ForTaskFunction taskFunctions, Class<?> forClass) {
+    return new TypedFunction(taskFunctions.getCollection(), forClass);
   }
 }

--- a/experimental/lambda/src/test/java/io/serverless/workflow/impl/executors/func/ForTaskFunctionRegressionTest.java
+++ b/experimental/lambda/src/test/java/io/serverless/workflow/impl/executors/func/ForTaskFunctionRegressionTest.java
@@ -17,10 +17,20 @@ package io.serverless.workflow.impl.executors.func;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import io.serverlessworkflow.api.types.Document;
+import io.serverlessworkflow.api.types.ForTaskConfiguration;
+import io.serverlessworkflow.api.types.Task;
+import io.serverlessworkflow.api.types.TaskItem;
 import io.serverlessworkflow.api.types.Workflow;
+import io.serverlessworkflow.api.types.func.CallJava;
+import io.serverlessworkflow.api.types.func.CallTaskJava;
 import io.serverlessworkflow.api.types.func.ForTaskFunction;
-import io.serverlessworkflow.fluent.func.FuncWorkflowBuilder;
 import io.serverlessworkflow.impl.WorkflowApplication;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.lang.reflect.Field;
 import java.util.Collection;
 import java.util.List;
 import java.util.concurrent.ExecutionException;
@@ -38,23 +48,75 @@ class ForTaskFunctionRegressionTest {
   }
 
   @Test
-  void forLoopWithoutExplicitCollectionClassExecutesSuccessfully()
+  void optionalFieldsSurviveJavaSerializationRoundTrip() throws Exception {
+    ForTaskFunction taskFunction = new ForTaskFunction();
+    clearField(taskFunction, "whileClass");
+    clearField(taskFunction, "itemClass");
+    clearField(taskFunction, "forClass");
+    clearField(taskFunction, "collection");
+
+    ForTaskFunction copy = roundTrip(taskFunction);
+
+    assertThat(copy.getWhileClass()).isNotNull().isEmpty();
+    assertThat(copy.getItemClass()).isNotNull().isEmpty();
+    assertThat(copy.getForClass()).isNotNull().isEmpty();
+  }
+
+  @Test
+  void forLoopWithExplicitCollectionClassExecutesSuccessfully()
       throws InterruptedException, ExecutionException {
     try (WorkflowApplication app = WorkflowApplication.builder().build()) {
+      ForTaskConfiguration forConfig = new ForTaskConfiguration();
       Workflow workflow =
-          FuncWorkflowBuilder.workflow("loop-without-collection-class")
-              .tasks(
-                  tasks ->
-                      tasks.forEach(
-                          f ->
-                              f.whileC(CallTest::isEven)
-                                  .collection(v -> (Collection<?>) v)
-                                  .tasks(CallTest::sum)))
-              .build();
+          new Workflow()
+              .withDocument(
+                  new Document()
+                      .withNamespace("test")
+                      .withName("loop-with-collection-class")
+                      .withVersion("1.0"))
+              .withDo(
+                  List.of(
+                      new TaskItem(
+                          "forLoop",
+                          new Task()
+                              .withForTask(
+                                  new ForTaskFunction()
+                                      .withWhile(CallTest::isEven)
+                                      .withCollection(v -> v, Collection.class)
+                                      .withFor(forConfig)
+                                      .withDo(
+                                          List.of(
+                                              new TaskItem(
+                                                  "javaCall",
+                                                  new Task()
+                                                      .withCallTask(
+                                                          new CallTaskJava(
+                                                              CallJava.loopFunction(
+                                                                  CallTest::sum,
+                                                                  forConfig.getEach()))))))))));
 
       var result = app.workflowDefinition(workflow).instance(List.of(2, 4, 6)).start().get();
 
       assertThat(result.asNumber().orElseThrow()).isEqualTo(12);
     }
+  }
+
+  private static ForTaskFunction roundTrip(ForTaskFunction taskFunction) throws Exception {
+    ByteArrayOutputStream output = new ByteArrayOutputStream();
+    try (ObjectOutputStream oos = new ObjectOutputStream(output)) {
+      oos.writeObject(taskFunction);
+    }
+
+    try (ObjectInputStream ois =
+        new ObjectInputStream(new ByteArrayInputStream(output.toByteArray()))) {
+      return (ForTaskFunction) ois.readObject();
+    }
+  }
+
+  private static void clearField(Object target, String fieldName)
+      throws ReflectiveOperationException {
+    Field field = target.getClass().getDeclaredField(fieldName);
+    field.setAccessible(true);
+    field.set(target, null);
   }
 }

--- a/experimental/lambda/src/test/java/io/serverless/workflow/impl/executors/func/ForTaskFunctionRegressionTest.java
+++ b/experimental/lambda/src/test/java/io/serverless/workflow/impl/executors/func/ForTaskFunctionRegressionTest.java
@@ -1,0 +1,60 @@
+/*
+ * Copyright 2020-Present The Serverless Workflow Specification Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.serverless.workflow.impl.executors.func;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.serverlessworkflow.api.types.Workflow;
+import io.serverlessworkflow.api.types.func.ForTaskFunction;
+import io.serverlessworkflow.fluent.func.FuncWorkflowBuilder;
+import io.serverlessworkflow.impl.WorkflowApplication;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import org.junit.jupiter.api.Test;
+
+class ForTaskFunctionRegressionTest {
+
+  @Test
+  void initializesOptionalFieldsAsEmpty() {
+    ForTaskFunction taskFunction = new ForTaskFunction();
+
+    assertThat(taskFunction.getWhileClass()).isNotNull().isEmpty();
+    assertThat(taskFunction.getItemClass()).isNotNull().isEmpty();
+    assertThat(taskFunction.getForClass()).isNotNull().isEmpty();
+  }
+
+  @Test
+  void forLoopWithoutExplicitCollectionClassExecutesSuccessfully()
+      throws InterruptedException, ExecutionException {
+    try (WorkflowApplication app = WorkflowApplication.builder().build()) {
+      Workflow workflow =
+          FuncWorkflowBuilder.workflow("loop-without-collection-class")
+              .tasks(
+                  tasks ->
+                      tasks.forEach(
+                          f ->
+                              f.whileC(CallTest::isEven)
+                                  .collection(v -> (Collection<?>) v)
+                                  .tasks(CallTest::sum)))
+              .build();
+
+      var result = app.workflowDefinition(workflow).instance(List.of(2, 4, 6)).start().get();
+
+      assertThat(result.asNumber().orElseThrow()).isEqualTo(12);
+    }
+  }
+}

--- a/experimental/types/src/main/java/io/serverlessworkflow/api/types/func/ForTaskFunction.java
+++ b/experimental/types/src/main/java/io/serverlessworkflow/api/types/func/ForTaskFunction.java
@@ -16,6 +16,8 @@
 package io.serverlessworkflow.api.types.func;
 
 import io.serverlessworkflow.api.types.ForTask;
+import java.io.IOException;
+import java.io.ObjectInputStream;
 import java.util.Collection;
 import java.util.Optional;
 import java.util.function.Function;
@@ -28,6 +30,10 @@ public class ForTaskFunction extends ForTask {
   private Optional<Class<?>> itemClass = Optional.empty();
   private Optional<Class<?>> forClass = Optional.empty();
   private Function<?, Collection<?>> collection;
+
+  public ForTaskFunction() {
+    normalizeOptionalFields();
+  }
 
   public <T, V> ForTaskFunction withWhile(LoopPredicate<T, V> whilePredicate) {
     return withWhile(toPredicate(whilePredicate));
@@ -112,8 +118,8 @@ public class ForTaskFunction extends ForTask {
       Optional<Class<?>> modelClass,
       Optional<Class<?>> itemClass) {
     this.whilePredicate = whilePredicate;
-    this.whileClass = modelClass;
-    this.itemClass = itemClass;
+    this.whileClass = modelClass != null ? modelClass : Optional.empty();
+    this.itemClass = itemClass != null ? itemClass : Optional.empty();
     return this;
   }
 
@@ -146,5 +152,22 @@ public class ForTaskFunction extends ForTask {
 
   public Function<?, Collection<?>> getCollection() {
     return collection;
+  }
+
+  private void normalizeOptionalFields() {
+    if (whileClass == null) {
+      whileClass = Optional.empty();
+    }
+    if (itemClass == null) {
+      itemClass = Optional.empty();
+    }
+    if (forClass == null) {
+      forClass = Optional.empty();
+    }
+  }
+
+  private void readObject(ObjectInputStream input) throws IOException, ClassNotFoundException {
+    input.defaultReadObject();
+    normalizeOptionalFields();
   }
 }

--- a/experimental/types/src/main/java/io/serverlessworkflow/api/types/func/ForTaskFunction.java
+++ b/experimental/types/src/main/java/io/serverlessworkflow/api/types/func/ForTaskFunction.java
@@ -31,10 +31,6 @@ public class ForTaskFunction extends ForTask {
   private Optional<Class<?>> forClass = Optional.empty();
   private Function<?, Collection<?>> collection;
 
-  public ForTaskFunction() {
-    normalizeOptionalFields();
-  }
-
   public <T, V> ForTaskFunction withWhile(LoopPredicate<T, V> whilePredicate) {
     return withWhile(toPredicate(whilePredicate));
   }
@@ -118,8 +114,8 @@ public class ForTaskFunction extends ForTask {
       Optional<Class<?>> modelClass,
       Optional<Class<?>> itemClass) {
     this.whilePredicate = whilePredicate;
-    this.whileClass = modelClass != null ? modelClass : Optional.empty();
-    this.itemClass = itemClass != null ? itemClass : Optional.empty();
+    this.whileClass = modelClass;
+    this.itemClass = itemClass;
     return this;
   }
 
@@ -168,6 +164,7 @@ public class ForTaskFunction extends ForTask {
 
   private void readObject(ObjectInputStream input) throws IOException, ClassNotFoundException {
     input.defaultReadObject();
+    // Preserve compatibility with older serialized instances that may have null optionals.
     normalizeOptionalFields();
   }
 }

--- a/experimental/types/src/main/java/io/serverlessworkflow/api/types/func/ForTaskFunction.java
+++ b/experimental/types/src/main/java/io/serverlessworkflow/api/types/func/ForTaskFunction.java
@@ -24,9 +24,9 @@ public class ForTaskFunction extends ForTask {
 
   private static final long serialVersionUID = 1L;
   private LoopPredicateIndexFilter<?, ?> whilePredicate;
-  private Optional<Class<?>> whileClass;
-  private Optional<Class<?>> itemClass;
-  private Optional<Class<?>> forClass;
+  private Optional<Class<?>> whileClass = Optional.empty();
+  private Optional<Class<?>> itemClass = Optional.empty();
+  private Optional<Class<?>> forClass = Optional.empty();
   private Function<?, Collection<?>> collection;
 
   public <T, V> ForTaskFunction withWhile(LoopPredicate<T, V> whilePredicate) {

